### PR TITLE
processes: add contextswitches a configurable option & use buffer to read /proc/stat 's content only once

### PR DESCRIPTION
**Pull request by zzzyhtheonly**
Changelog: add contextswitches a configurable option & use buffer to read /proc/stat 's content only once

We already have a plugin "contextswitch" to collect "ctxt" in "/proc/stat" in Linux. But since we also read from "/proc/stat" from plugin "processes". We add an option to enable "contextswitch" in plugin "processes", default disabled. For this feature, please see 58a1e09e88f2cce1640a4072f7527c54832901d6 

Also, we find that we open and read "/proc/stat" repeatedly when we read "procs_running" and "fork_rate", so we also patch this by only reading "/proc/stat" once into a buffer and making "procs_running", "fork_rate" and "contextswitch" read from this buffer. For this feature, please see de2d033ba5ccd1531caab2e3772d471c219671eb

Tested on my local Linux machines; please help review. Thanks a lot!

### CODE REVIEW

File: src/processes.c
Comment (vincele) [inline src/processes.c:1586]:
```suggestion
    WARNING("'processes ' not found in /proc/stat");
```
Maybe a bit uglier, but conveys more info.

If you think it is better like that, there are other occurrences of the same kind of warning.

File: src/processes.c
Comment (vincele) [inline src/processes.c:1625]: Naive question: why not exit earlier if the value is not to be reported ?

File: src/processes.c
Comment (vincele) [inline src/processes.c:2155]:
```suggestion
    ERROR("Cannot read `/proc/stat`");
```

File: src/processes.c
Comment (vincele) [inline src/processes.c:2179]: maybe this is where the `if (report_sys_ctxt_switch)` should be ?

**Review by vincele — COMMENTED** [review]
LGTM

File: src/processes.c
Comment (vincele) [reply src/processes.c:1625]: see my other comment below...

**Comment by zzzyhtheonly** [comment]
@vincele Thank you! All resolved, please help review again.

File: src/collectd.conf.pod
Comment (eero-t) [inline src/collectd.conf.pod:8103]: Simpler as:
```suggestion
Can be configured only outside the B<Process> and B<ProcessMatch>
```

File: src/processes.c
Comment (eero-t) [inline src/processes.c:1618]: Seems odd to hard code value on other line.  Maybe do it like this (untested):
```suggestion
  const int expected = STATIC_ARRAY_SIZE(fields);
  fields_num = strsplit(ctxt, fields, expected);
  if (fields_num != expected)
```

File: src/processes.c
Comment (vincele) [inline src/processes.c:1472]:
```suggestion
    WARNING("'procs_running ' not found in /proc/stat");
```

File: src/processes.c
Comment (vincele) [inline src/processes.c:2182]: Is this (commented `#endif`) useful ?

**Review by vincele — CHANGES_REQUESTED** [review]
I think this still needs some small fixes

File: src/processes.c
Comment (vincele) [inline src/processes.c:983]: Why is that comment needed ?

File: src/processes.c
Comment (zzzyhtheonly) [reply src/processes.c:2179]: I guess this makes it clearer where the end of the "KERNEL_LINUX" part is. Since this kind of comment exists in other places, e.g., #319, #786, maybe we should keep it for this PR?

BTW, the starting spaces are caused by format.sh

File: src/processes.c
Comment (zzzyhtheonly) [reply src/processes.c]: Removed

File: src/processes.c
Comment (zzzyhtheonly) [reply src/processes.c]: Sorry for my carelessness, fixed.

**Review by vincele — COMMENTED** [review]
This diff has become hard to read, but it LGTM...

File: src/processes.c
Comment (eero-t) [inline src/processes.c:2181]: Does "format.sh" work better if that comment is associated with the code it refers to?
```suggestion
    read_sys_ctxt_switch(buffer);

  /* #endif KERNEL_LINUX */
#elif HAVE_LIBKVM_GETPROCS && HAVE_STRUCT_KINFO_PROC_FREEBSD
```

PS. this this is not the only place where such comments are wrongly associated & indented with the last code line, instead of the #elif/#else/#endif that actually terminates the preceeding #if/#elif/#else block.

Also, IMHO having `#endif` in the comments is misleading as it looks like commented-code instead of just a comment.  Esing e.g. "KERNEL_LINUX ends" would be clearer.

Comment (zsfz) [reply src/processes.c:2181]: I agree with your point, but it seems like this annotation style is prevalent in many plugins such as cpu plugin. Should we strive for consistency?

Comment (eero-t) [reply src/processes.c:2181]: IMHO:
* If that fixes the indenting done by `format.sh`, that would be nice, but that should be a separate PR, and some comment from current project maintainers would be nice too
* Above KERNEL_LINUX "define end" comment does not need to be changed in this PR, as it's how rests of the plugin code is written currently

Comment (zzzyhtheonly) [reply src/processes.c:2181]: Agreed, so for now I'll leave this comment unchanged here for this PR.

**Comment by eero-t** [comment]
@mrunge review feedback so far has been handled, do you want something still to be done before you merge?

**Comment by mrunge** [comment]
thank you, this looks good to me, and thank you for the effort following up with this.

**Pull request merged**
